src/linking/remove_internal_symbols.cpp

```diff
@@ -182,6 +183,8 @@ void remove_internal_symbols(
       {
         get_symbols(ns, symbol, exported);
       }
+      else if(is_contract)
+        get_symbols(ns, symbol, exported);
```

Comment:
If a function `foo` is file_local, is also `contract::foo` file_local ? 

What if some of the contract's clauses use file_local symbols in preconditions or post conditions or assigns clause ?

Reply:
No, the contract does not currently have any flags (other than `is_property` to mark it as a contract) set, and will therefore not be considered "file_local."

Your call-out about file_local symbols is a good one as the linker could choose to rename them in case of a conflict, at which point the contract would no longer refer to the correct symbol. I'll try to come up with some code that tracks this renaming and adjusts the contract accordingly.

Reply:
Hi again,
I experimented with the following:


```c
// foo.h
#include <stdlib.h>
void foo(char *arr, size_t size);
```

```c
// foo.c
#include "foo.h"
void foo(char *arr, size_t size) {
    if(arr && size > 0)
        arr[0] = 0;
}
```

```c
// foo_contract.c
#include <stdlib.h>
void foo(char *arr, size_t size)
__CPROVER_assigns(arr && size > 0 : arr[0]);
```

```c
// main.c
#include "foo.h"
void main() {
    size_t size = 10;
    char *arr = malloc(size);
    foo(arr, size);
}
```

and then 

```bash
▶goto-cc foo_contract.c foo.c main.c  -o project.o && goto-instrument project.o project-mod.o --enforce-contract foo && cbmc project-mod.o

** Results:
<builtin-library-malloc> function malloc
[malloc.assertion.1] line 26 max allocation size exceeded: SUCCESS
[malloc.assertion.2] line 31 max allocation may fail: SUCCESS

foo.c function foo
[foo.assigns.2] line 4 Check that arr[(signed long int)0] is assignable: SUCCESS
[foo.assigns.1] file foo_contract.c line 4 Check that arr[(signed long int)0] is valid when arr != ((char *)NULL) && size > (unsigned long int)0: SUCCESS

** 0 of 4 failed (1 iterations)
VERIFICATION SUCCESSFUL
```

But if I include `"foo.h"` in `"foo_contract.h"` like so

```c
// foo_contract.c
#include <stdlib.h>
#include <foo.h>
void foo(char *arr, size_t size)
__CPROVER_assigns(arr && size > 0 : arr[0]);
```


Then it seems like the contract clauses are lost.

```bash
** Results:
<builtin-library-malloc> function malloc
[malloc.assertion.1] line 26 max allocation size exceeded: SUCCESS
[malloc.assertion.2] line 31 max allocation may fail: SUCCESS

foo.c function foo
[foo.assigns.1] line 4 Check that arr[(signed long int)0] is assignable: FAILURE

** 1 of 3 failed (2 iterations)
VERIFICATION FAILED
```

Reply:
Good catch, we need to maintain the contract in `typecheck_redefinition_non_type` even when the new symbol does not provide the function definition. I'll add this case and will try to include a test.

Reply:
Thanks @remi-delmas-3000, I believe I have now fixed this problem (I have also extended the test suite to capture this scenario).